### PR TITLE
KOGITO-3913: Integration tests for Standalone DMN Editor

### PR DESCRIPTION
**Pull request by jstastny-cz**

### CODE REVIEW

**Comment by jstastny-cz** [comment]
@tiagobento @paulovmr @tomasdavidorg Could you please review the build approach?

Justification:
* For standalone editors integration testing I want to have a webapp serving various configs of editors on various pages
  * Created reusable editor react wrapper.
  * Added `tests-it` package with the app (the app will be tested using cypress).
* During the build Node runs out of memory when either:
  * Whole project built as `build:fast` - then `tests-it` build runs out of memory (bundles of editors have >100MB each).
  * Whole project built as `build:prod` - and also `tests-it` as `build:prod` - runs out of memory (due to webpack --mode production)
  * Works only when - all is built as `build:prod`, but the `tests-it` without `--mode production` (i.e. like `build:fast`).
* As a result I've extracted the build of `tests-it` package by not providing the `build:prod` and `build:fast` scripts.
* Introduced a `build:it` script that does what `build:fast` (due to separation from the build).
* In it-tests GH workflow I've introduced a separate job for this new Integration tests - see the yaml file changes.
  * Needed a separate job, cause the other one requires `build:fast` which does not fit the needs of the new suite (reasoning above).
  * Using lerna's selective build to optimize the time (can be also applied to the other job there).
  * This could be the encouraged approach for all jobs there - build just what's necessary the way you need (prod vs fast).

**Comment by tiagobento** [comment]
@jstastny-cz Looks like GH Actions got completely lost on this PR D: I believe if you send a new commit things will get back to normal.

**Comment by jstastny-cz** [comment]
@tiagobento I've included the it-tests under the package itself, but I failed at configuring the editor bundles as externals. Is that something you can help me with?

**Comment by tiagobento** [comment]
@jstastny-cz In order to have your externals configured properly, you need to add a script tag pointing to the result of the build of `kie-editors-standalone`, that means you need to alter your webserver to serve files from there as well. Also, you need to add an `externals` property on your `it-tests/webpack.config.js` file with the name of the module you want to externalize. In this case `@kogito-tooling/kie-editors-standalone`.

File: packages/kie-editors-standalone/it-tests/cypress/support/commands.d.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    editor(id: string, options?: Record<string, any>): Chainable<Element>;
+
+    /**
+     * Upload file using elements with type file-loader and given id.
+     * @param fileName string name of the file
+     * @param componentId string id of the element
+     * @param options
+     */
+    uploadFile(fileName: string, componentId: string, options?: Record<string, any>): Chainable<Element>;
```
Comment (jstastny-cz) [inline packages/kie-editors-standalone/it-tests/cypress/support/commands.d.ts]: not correct - need to have just single command with the name.

File: packages/kie-editors-standalone/it-tests/cypress/support/commands.d.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    uploadFile(fileName: string, options?: Record<string, any>): Chainable<Element>;
+
+    /**
+     * Load file contents into editor by clicking the View button near the file-list item.
+     * @param fileName string name of the file
+     * @param componentId string id of the element
+     * @param options
+     */
+    viewFile(fileName: string, componentId: string, options?: Record<string, any>): Chainable<Element>;
```
Comment (jstastny-cz) [inline packages/kie-editors-standalone/it-tests/cypress/support/commands.d.ts]: again, need to remove one declaration

**Comment by jstastny-cz** [comment]
Failing due to KOGITO-3905 and KOGITO-3906 being fixed in https://github.com/kiegroup/kie-wb-common/pull/3512

File: packages/kie-editors-standalone/it-tests/cypress.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "fixturesFolder": "it-tests/cypress/fixtures",
```
Comment (tomasdavidorg) [inline packages/kie-editors-standalone/it-tests/cypress.json:2]: Is the `cypress` directory needed? Would it be possible to have just `"it-tests/fixtures"`, `"it-tests/integration"`, ...?

File: packages/kie-editors-standalone/it-tests/.gitignore [hunk 4]
```diff
@@ -0,0 +1,7 @@
+node_modules/
+dist/
+coverage/
+target/
+.idea
+cypress/screenshots
```
Comment (tomasdavidorg) [inline packages/kie-editors-standalone/it-tests/.gitignore]: Do we need this .gitignore file? The package `kie-editors-standalone` already have one.

File: packages/kie-editors-standalone/it-tests/cypress/integration/both-bpmn-dmn.spec.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+describe("Both BPMN DMN.", () => {
+  before("Visit page", () => {
+    cy.visit("localhost:9001/both-bpmn-dmn");
+    cy.loadEditor(["both-bpmn", "both-dmn"]);
+  });
+  it("Test Load File And View", () => {
```
Comment (tomasdavidorg) [inline packages/kie-editors-standalone/it-tests/cypress/integration/both-bpmn-dmn.spec.ts:23]: I would prefer empty line before the test.

File: packages/kie-editors-standalone/it-tests/cypress.json (hunk 3 above)
Comment (jstastny-cz) [reply packages/kie-editors-standalone/it-tests/cypress.json:2]: sure, it is possible, I thought I would distinguish more source folders of the webapp used for testing and the actual tests. But I can flatten the structure. So resulting tree would be 
```
it-test
- src / - the webapp sources
- fixtures / 
- plugins /
- integration / 
- support /
```

File: packages/kie-editors-standalone/it-tests/.gitignore (hunk 4 above)
Comment (jstastny-cz) [reply packages/kie-editors-standalone/it-tests/.gitignore]: most probably not, will remove.

File: packages/kie-editors-standalone/it-tests/cypress/integration/both-bpmn-dmn.spec.ts (hunk 5 above)
Comment (jstastny-cz) [reply packages/kie-editors-standalone/it-tests/cypress/integration/both-bpmn-dmn.spec.ts:23]: ok, will reformat.

File: packages/kie-editors-standalone/it-tests/cypress.json (hunk 3 above)
Comment (tomasdavidorg) [reply packages/kie-editors-standalone/it-tests/cypress.json:2]: Ok lets keep the cypress dir. Thanks

Comment (jstastny-cz) [reply packages/kie-editors-standalone/it-tests/cypress.json:2]: thank you.

**Comment by tiagobento** [comment]
@jstastny-cz Do you mind pushing an empty commit to this PR? We fixed our CI :: Full downsteram check and a new commit will trigger it with the fixed script.

`git commit --amend --no-edit && git push -f [remote] [branch]`

**Comment by jstastny-cz** [comment]
Summary:

- all: dmn-read-only tests failing due to fix from kie-wb-common not being yet promoted to kogito-tooling (needs version bump or run as part of FDB).
- mac-os: seeing more "stable" failures on the env - need to investigate on a mac (any help is welcome).

**Comment by tiagobento** [comment]
@jstastny-cz As we spoke the other day, we need to add these tests to our `CI :: Full Downstream build` as well, right? If you do that in this PR, we won't be seeing the results, since this job is always fetched from `master`. But if we do that on a separate PR and merge it, `CI :: Full Downstream build` will be broken for the other PRs :D

My suggestion is for you to do the changes on `CI :: Full Downstream build` on the `master` branch of your fork, open a PR much like this one, but targeting some branch on your fork, and watch the result of the Workflow. If it's successful, simply apply the changes you've made on `CI :: Full Downstream build` to this branch, and we can merge it. If you can think of a better way of doing all this, please share!

Thanks!

**Comment by jstastny-cz** [comment]
@tiagobento thanks for the hints.

Currently I am looking into macOS failures. For getting this PR merged (so that it does not break anything) the easiest way is to temporarily disable the failing test and add only after the dependency is updated.

Adding into FDB could be a next step (in a smaller PR), wdyt?
Course of action would be:
1. remove the failing test for dmn read-only mode
2. fix tests on macos
3. get this merged
-----
4. open a PR where adding the test execution also to FDB

**Comment by tiagobento** [comment]
@jstastny-cz Ok, sounds good! :)

**Comment by jstastny-cz** [comment]
Ok, so macos build can actually be green, will try a new emtpy commit to trigger actions again.

**Comment by jstastny-cz** [comment]
last run - just ubuntu failed on long `Loading...` spinner showing (more than 60 s, probably just a hickup)

**Comment by jstastny-cz** [comment]
ok, it-tests are all green (but now failures in Default build :/ )

**Comment by tiagobento** [comment]
@jstastny-cz Yeah...... macOS has that issue. I still don't know what causes it and my researches never give me something significant. Anyway. I'll just re-run it and it'll pass :)

**Review by tiagobento — APPROVED** [review]

**Review by tomasdavidorg — APPROVED** [review]

**Review by paulovmr — APPROVED** [review]
Looks good, thanks @jstastny-cz !

**Pull request merged**